### ISSUE TITLE
the covariance matrix in the channel dimension

### ISSUE THREAD

**Issue by zAuk000** [Kunhao-Liu/StyleRF#1]
Hi, thanks for sharing your work ! and I have some questions about the code, there is a  covariance matrix Cov(Q,K) in the channel dimension in section 3.2.1 , but I noticed that the code about the matrix is c_cov = q_embed.transpose(1,2).unsqueeze(3) * k_embed.transpose(1,2).unsqueeze(2) in styleModules.py that seems like outer product rather than Cov, do I locate the wrong code or make some misunderstandings?

**Comment by zAuk000** [comment]
Hi，
Did you use the entire WikiArt dataset as the training set? I used all the photos as the training set, but I didn't get the desired performance from the pretrained model. Could it be a problem with my parameter settings?

**Comment by Kunhao-Liu** [comment]
Hi, sorry for the late reply. For the covariance matrix, we follow the implementation of [Linear Style Transfer](https://github.com/sunshineatnoon/LinearStyleTransfer), you can check their interpretation. 

We use the entire WikiArt dataset for training. If you get unwanted results (e.g. the content or style is not well preserved), you can change `content_weight` or `style_weight`  to balance the content and style preservation. Also, `patch_size` should not be too small, otherwise the results may also degrade.

**Comment by zAuk000** [comment]
> Hi, sorry for the late reply. For the covariance matrix, we follow the implementation of [Linear Style Transfer](https://github.com/sunshineatnoon/LinearStyleTransfer), you can check their interpretation.
> 
> We use the entire WikiArt dataset for training. If you get unwanted results (e.g. the content or style is not well preserved), you can change `content_weight` or `style_weight` to balance the content and style preservation. Also, `patch_size` should not be too small, otherwise the results may also degrade.

Thank you very much for your reply! Okay, I will try adjusting the weights. Previously, I used the original parameters provided in the llff_style.txt file for training. Perhaps you also keep the corresponding parameters for the trex pre-trained model？Could I request a copy？Thank you.

**Comment by Kunhao-Liu** [comment]
You can find the pre-trained checkpoints at [this url](https://drive.google.com/drive/folders/1nF9-6lTIhktG5JjNvnmdYOo1LTvtK7Dw?usp=share_link). Thanks.

**Comment by zAuk000** [comment]
> 

Yes, thank you very much for your sharing. In fact, I want to train from scratch and  train without modifying the parameters in the llff_style.txt. However, I found that the performance of the obtained model is not as good as the provided pre-trained model. So, I would like to know if the training parameters I used in the llff_style.txt are different from the ones you used for pre-trained model, or if I should try modifying the weights and train again? Thank you.

**Comment by Kunhao-Liu** [comment]
It's strange that you're encountering this issue. We used the same training parameters as specified in the llff_style.txt file. Have you tried testing the same reference style images as the ones demonstrated on the project page? Alternatively, you could train the model for additional iterations (for instance, setting `n_iters` to 50k) to see if the results improve.

**Comment by zAuk000** [comment]
Yes, considering the potential differences in style data selection, I have tried increasing the training iterations. However, even with up to fifty or even a hundred thousand iterations, the performance of the final model remains roughly the same. I have also considered the possibility of style image issues during inference and have tested several other images. The conclusion drawn is that the entire model hasn't been trained properly. For instance, the following are comparisons of the same style image. Furthermore, I attempted to retrain the model from scratch, which resulted in slight improvements but still fell short of expectations. I am using an RTX 3090 graphics card. Perhaps I would like to know if the pretraining models in the first two steps would have a significant impact on the results of the third step？Thank you very muck for your patience！
![001](https://github.com/Kunhao-Liu/StyleRF/assets/97910692/ad8cd8bf-4645-47b1-b406-6c62ad68fdc3)
![006](https://github.com/Kunhao-Liu/StyleRF/assets/97910692/bc296527-f309-4267-bf5b-f60b4c20186c)

**Comment by zAuk000** [comment]
Yes, considering the potential differences in style data selection, I have tried increasing the training iterations. However, even with up to fifty or even a hundred thousand iterations, the performance of the final model remains roughly the same. I have also considered the possibility of style image issues during inference and have tested several other images. The conclusion drawn is that the entire model hasn't been trained properly. For instance, the following are comparisons of the same style image. Furthermore, I attempted to retrain the model from scratch, which resulted in slight improvements but still fell short of expectations. I am using an RTX 3090 graphics card. Perhaps I would like to know if the pretraining models in the first two steps would have a significant impact on the results of the third step.Thank you very much for your patience！
![001(1)](https://github.com/Kunhao-Liu/StyleRF/assets/97910692/c24de31a-176f-4be9-b11b-76c4a5859b23)
![006(1)](https://github.com/Kunhao-Liu/StyleRF/assets/97910692/b8359789-190c-4bbf-a556-7047e5edfe38)

**Comment by Kunhao-Liu** [comment]
Hi, I believe the first two steps of training might not have a significant impact on the quality of stylization. Based on your results, the content preservation seems quite poor. Perhaps you could consider setting the `style_weight` to a smaller value, such as 10?

**Comment by zAuk000** [comment]
In fact, I have already tried modifying the corresponding weights, but the difference is still not significant. I feel that the rendered images from the trained model tend to be darker, which does not align with the style of the style images. It is worth noting that there are significant differences in the feature maps obtained from the log_style file. Due to my lack of knowledge in style transfer, I am not sure where the differences in feature map extraction might be coming from.I think this is indeed related to content preservation. I will try modifying the style weight to a smaller value, such as 2. However, when using the original parameters, the resulting model's performance differs significantly, and I'm not sure where the issue lies.

![056(1)](https://github.com/Kunhao-Liu/StyleRF/assets/97910692/389d8bd1-f072-428a-8e84-b3cd35fc2738)![feature_058](https://github.com/Kunhao-Liu/StyleRF/assets/97910692/2ee39e82-29dd-428a-a695-541fcebe38a2)

![000(1)](https://github.com/Kunhao-Liu/StyleRF/assets/97910692/1c248ed8-2e87-4a27-b775-186453ff835e)![feature_000(1)](https://github.com/Kunhao-Liu/StyleRF/assets/97910692/6f172232-a116-430a-bb60-2f13a1ce3bb7)

**Comment by zAuk000** [comment]
Hi，I think I've found the problem. In the old version of torchvision（0.11.0）, the keyword "weights=DEFAULT" cannot be used. Use "Pretrained=True" instead. When I first encountered this error, I thought I had replaced it with "Pretrained=True", but in reality I hadn't. So the problem was that I didn't load the pretrained VGG. Now, after several training iterations, I have achieved similar visual effects to the pretrained model, although not identical. Once again, thank you for your patience in answering and I apologize for bothering you multiple times.
